src/camera_parameters.cpp

```diff
@@ -0,0 +1,601 @@
+    }
+    return true;
+  } catch (std::runtime_error e) {
+    ROS_ERROR("%s", e.what());
+    return false;
+  }
+}
+
+bool CameraParametersPair::setInputCameraParameters(
+    const cv::Size& resolution, const Eigen::Matrix<double, 4, 4>& T,
+    const Eigen::Matrix<double, 3, 3>& K, const std::vector<double>& D,
+    const bool radtan_distortion) {
+  try {
+    input_ptr_ = std::make_shared<InputCameraParameters>(resolution, T, K, D,
+                                                         radtan_distortion);
+    return true;
+  } catch (std::runtime_error e) {
+    ROS_ERROR("%s", e.what());
+    return false;
+  }
+}
+
+bool CameraParametersPair::setOutputCameraParameters(
+    const cv::Size& resolution, const Eigen::Matrix<double, 4, 4>& T,
+    const Eigen::Matrix<double, 3, 3>& K) {
+  try {
+    output_ptr_ = std::make_shared<OutputCameraParameters>(resolution, T, K);
+    return true;
+  } catch (std::runtime_error e) {
+    ROS_ERROR("%s", e.what());
+    return false;
+  }
+}
+
+bool CameraParametersPair::setOutputFromInput() {
+  if (!valid(true)) {
+    ROS_ERROR(
+        "Cannot set output to same values as input, as input is not currently "
+        "set");
+    return false;
+  } else {
+    setOutputCameraParameters(input_ptr_->resolution(), input_ptr_->T(),
+                              input_ptr_->K());
+    return true;
+  }
+}
+
+bool CameraParametersPair::setOptimalOutputCameraParameters(
+    const double scale) {
+  if (!valid(true)) {
+    ROS_ERROR(
+        "Optimal output camera parameters cannot be set until the input camera "
+        "parameters have been given");
+    return false;
+  }
+  cv::Size resolution_estimate(input_ptr_->resolution().width,
+                               input_ptr_->resolution().height);
+  double focal_length =
+      scale * (input_ptr_->K()(0, 0) + input_ptr_->K()(1, 1)) / 2;
+  Eigen::Matrix<double, 3, 4> P = Eigen::Matrix<double, 3, 4>::Zero();
+  P(0, 0) = focal_length;
+  P(1, 1) = focal_length;
+  P(2, 2) = 1;
+  P(0, 2) = static_cast<double>(resolution_estimate.width) / 2.0;
+  P(1, 2) = static_cast<double>(resolution_estimate.height) / 2.0;
+  P.topRightCorner<3, 1>() = focal_length * input_ptr_->p();
+
+  std::vector<double> D;
+  if (undistort_) {
+    D = input_ptr_->D();
+  } else {
+    D = std::vector<double>(0, 5);
+  }
+
+  // Find the resolution of the output image
+  // Thanks to weird corner cases this is way more complex then it should be.
+  // The general case is even more of a nightmare, so we constrain the problem
+  // such that the center of focus must be in the center of the final image.
+
+  // as we are missing the forward projection model we iteratively estimate
+  // image size assuming a linear relationship between warping and size at
+  // each
+  // step
+  for (size_t i = 0; i < kFocalLengthEstimationAttempts; ++i) {
+    // get list of edge points to check
+    std::vector<Eigen::Vector2d, Eigen::aligned_allocator<Eigen::Vector2d>>
```

Comment:
The undistorter needs these values as doubles as it makes no assumption about points coming from a 2d grid.